### PR TITLE
config: reformat and add optional/required

### PR DESCRIPTION
**Pull request by philips**
Reformat this file to look more like manifest.md and add
OPTIONAL/REQUIRED tags on all fields.

Fixes https://github.com/opencontainers/image-spec/issues/251

### CODE REVIEW

File: config.md [hunk 1]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+     Default arguments to the entrypoint of the container.
+     These values act as defaults and are replaced with any specified when creating a container.
+     If an `Entrypoint` value is not specified, then the first entry of the `Cmd` array should be interpreted as the executable to run.
+
+   - **Volumes** *object*, OPTIONAL
+     A set of directories which should be created as data volumes in a container running this image. This field MAY be "null".
+     If a file or folder exists within the image with the same path as a data volume, that file or folder is replaced with the data volume and is never merged. **NOTE:** This JSON structure value is unusual because it is a direct JSON serialization of the Go type <code>map[string]struct{}</code> and is represented in JSON as an object mapping its keys to an empty object.
+
+   - **WorkingDir** *string*, REQUIRED
```
Comment (coolljt0725) [inline config.md]: why the `config` is OPTIONAL, but `WorkingDir` is REQUIRED? I think all the filed in `config` is OPTIONAL  :)

File: config.md [hunk 2]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+  - **created_by** *string*, OPTIONAL
+
+     The command which created the layer.
+
+  - **comment** *string*, OPTIONAL
+
+     A custom message set when creating the layer.
+
+  - **empty_layer** *string*
```
Comment (coolljt0725) [inline config.md]: I think this is a `Boolean`, not a `string`

File: config.md (hunk 1 above)
Comment (coolljt0725) [inline config.md]: Does this mean if `config` is not null, then the `WorkingDir` is `REQUIRED`?

**Comment by wking** [comment]
On Tue, Sep 20, 2016 at 08:36:59PM -0700, Lei Jitang wrote:
> +   - **WorkingDir** *string*, REQUIRED
>
> why the `config` is OPTIONAL, but `WorkingDir` is REQUIRED? I think
> all the filed in `config` is OPTIONAL :)

This means that if you set ‘config’ you MUST set ‘config.WorkingDir’.
And it's probably required because the runtime-spec requires
process.cwd [1].  I've filed opencontainers/runtime-spec#489 to relax
the process.args requirement, and if that ever gets legs I may file a
follow-up relaxing process.cwd.  But until the runtime-spec relaxes
the requirement (possibly never), I think image-spec is best off
reinventing as little of the wheel as possible ;).

[1]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0-rc2/config.md#process-configuration

File: config.md [hunk 3]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
-    "created": "2015-10-31T22:22:55.613815829Z",
-    "created_by": "/bin/sh -c #(nop) CMD [\"sh\"]",
-    "empty_layer": true
-  }
-]</pre>
-    </dd>
-</dl>
+- **created** *string*, OPTIONAL
+
+  A ISO-8601 formatted combined date and time at which the image was created.
```
Comment (wking) [inline config.md]: “A ISO” → “An ISO”.

File: config.md [hunk 4]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+  The name of the operating system which the image is built to run on.
+  Possible values include: `darwin`, `freebsd`, `linux`.
+  More values may be supported in the future and any of these may or may not be supported by a given container runtime implementation.
+  New entries SHOULD be submitted to this specification for standardization and be inspired by the [Go language documentation for $GOOS and $GOARCH](https://golang.org/doc/install/source#environment).
+
+- **config** *object*, OPTIONAL
+
+  The execution parameters which should be used as a base when running a container using the image.
+  This field can be <code>null</code>, in which case any execution parameters should be specified at creation of the container.
```
Comment (wking) [inline config.md]: `<code>` → backticks.

File: config.md [hunk 5]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+   - **user** *string*, OPTIONAL
```
Comment (wking) [inline config.md]: This was previously title-case, and as of 403a1af `specs-go/v1/config.go` still has:

    User string `json:"User"`

File: config.md [hunk 6]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+     The username or UID which the process in the container should run as.
+     This acts as a default value to use when the value is not specified when creating a container.
+     All of the following are valid: `user`, `uid`, `user:group`, `uid:gid`, `uid:group`, `uiser:gid`
```
Comment (wking) [inline config.md]: `uiser:gid` → `user:gid`.

File: config.md [hunk 7]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+The JSON structure also references a cryptographic hash of each layer used by the image, and provides history information for those layers.
+This JSON is considered to be immutable, because changing it would change the computed ImageID.
+Changing it means creating a new derived image, instead of changing the existing image.
+
+### Layer DiffID
+
+A layer DiffID is a SHA256 digest over the layer's uncompressed tar archive and serialized in the descriptor digest format, e.g., `sha256:a9561eb1b190625c9adb5a9513e72c4dedafc1cb2d4c5236c9a6957ec7dfd5a9`.
+Layers must be packed and unpacked reproducibly to avoid changing the layer DiffID, for example by using tar-split to save the tar headers.
+
+NOTE: the DiffID is different than the digest in the manifest list because the manifest digest is taken over the gzipped layer for <code>application/vnd.oci.image.layer.tar+gzip</code> types.
```
Comment (wking) [inline config.md]: `<code>` → backticks.

File: config.md [hunk 8]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+     If `group`/`gid` is not specified, the default group and supplementary groups of the given `user`/`uid` in `/etc/passwd` from the container are applied.
+
+   - **Memory** *integer*, OPTIONAL
+
+     Memory limit (in bytes).
+     This acts as a default value to use when the value is not specified when creating a container.
+
+   - **MemorySwap** *integer*, OPTIONAL
+
+     Total memory usage (memory + swap); set to <code>-1</code> to disable swap.
```
Comment (wking) [inline config.md]: `<code>` → backticks.

File: config.md [hunk 9]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+     CPU shares (relative weight vs. other containers).
+     This acts as a default value to use when the value is not specified when creating a container.
+
+   - **ExposedPorts** *object*, OPTIONAL
+
+     A set of ports to expose from a container running this image.
+     Its keys can be in the format of:
+`port/tcp`, `port/udp`, `port` with the default protocol being `tcp` if not specified.
+     These values act as defaults and are merged with any specified when creating a container.
+     **NOTE:** This JSON structure value is unusual because it is a direct JSON serialization of the Go type <code>map[string]struct{}</code> and is represented in JSON as an object mapping its keys to an empty object.
```
Comment (wking) [inline config.md]: `<code>` → backticks.

File: config.md [hunk 10]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+   - **Env** *array of strings*, OPTIONAL
+
+     Entries are in the format of <code>VARNAME="var value"</code>.
```
Comment (wking) [inline config.md]: `<code>` → backticks.

File: config.md [hunk 11]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+     These values act as defaults and are merged with any specified when creating a container.
+
+   - **Entrypoint** *array of strings*
+
+     A list of arguments to use as the command to execute when the container starts.
+     This value acts as a  default and is replaced by an entrypoint specified when creating a container. This field MAY be "null".
```
Comment (wking) [inline config.md]: `"null"` → `null`?

Comment (wking) [inline config.md]: Also remove the doubled space: “as a  default“ → “as a default”.  And move “This field MAY be `null`” to its own line.

File: config.md [hunk 12]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+   - **Cmd** *array of strings*, OPTIONAL
+
+     Default arguments to the entrypoint of the container.
+     These values act as defaults and are replaced with any specified when creating a container.
+     If an `Entrypoint` value is not specified, then the first entry of the `Cmd` array should be interpreted as the executable to run.
+
+   - **Volumes** *object*, OPTIONAL
+     A set of directories which should be created as data volumes in a container running this image. This field MAY be "null".
+     If a file or folder exists within the image with the same path as a data volume, that file or folder is replaced with the data volume and is never merged. **NOTE:** This JSON structure value is unusual because it is a direct JSON serialization of the Go type <code>map[string]struct{}</code> and is represented in JSON as an object mapping its keys to an empty object.
```
Comment (wking) [inline config.md]: `<code>` → backticks.

File: config.md [hunk 13]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+  - **diff_ids** *array*, REQUIRED
+
+     An array of layer content hashes (`DiffIDs`), in order from bottom-most to top-most.
+
+- **history** *array of object*
+
+ Describes the history of each layer.
+ The array is ordered from bottom-most layer to top-most layer.
+ The object has the following fields:
```
Comment (wking) [inline config.md]: Missing indents?

File: config.md [hunk 14]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+  - **created** *string*, OPTIONAL
```
Comment (wking) [inline config.md]: Four-space indents for sub-lists (opencontainers/runtime-spec#495).

File: config.md [hunk 15]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+
+     The command which created the layer.
+
+  - **comment** *string*, OPTIONAL
+
+     A custom message set when creating the layer.
+
+  - **empty_layer** *string*
+
+     This field is used to mark if the history item created a filesystem diff, OPTIONAL
```
Comment (wking) [inline config.md]: “OPTIONAL“ should go on the line with **empty_layer**, like you do with all the other properties.

File: config.md [hunk 16]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+   - **Volumes** *object*, OPTIONAL
+     A set of directories which should be created as data volumes in a container running this image. This field MAY be "null".
+     If a file or folder exists within the image with the same path as a data volume, that file or folder is replaced with the data volume and is never merged. **NOTE:** This JSON structure value is unusual because it is a direct JSON serialization of the Go type <code>map[string]struct{}</code> and is represented in JSON as an object mapping its keys to an empty object.
+
+   - **WorkingDir** *string*, REQUIRED
+
+     Sets the current working directory of the entrypoint process in the container.
+     This value acts as a default and is replaced by a working directory specified when creating a container.
+
+- **rootfs** *object, REQUIRED
```
Comment (coolljt0725) [inline config.md]: `*object` -> `*object*`

File: config.md [hunk 17]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+     This value acts as a  default and is replaced by an entrypoint specified when creating a container. This field MAY be "null".
+
+   - **Cmd** *array of strings*, OPTIONAL
+
+     Default arguments to the entrypoint of the container.
+     These values act as defaults and are replaced with any specified when creating a container.
+     If an `Entrypoint` value is not specified, then the first entry of the `Cmd` array should be interpreted as the executable to run.
+
+   - **Volumes** *object*, OPTIONAL
+     A set of directories which should be created as data volumes in a container running this image. This field MAY be "null".
```
Comment (coolljt0725) [inline config.md]: should have a blang line after `**Volumes** *object*, OPTIONAL`

File: config.md [hunk 18]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+Image filesystems are composed of *layers*.
+Each layer represents a set of filesystem changes in a tar-based [layer format](layer.md), recording files to be added, changed, or deleted relative to its parent layer.
+Layers do not have configuration metadata such as environment variables or default arguments - these are properties of the image as a whole rather than any particular layer.
+Using a layer-based or union filesystem such as AUFS, or by computing the diff from filesystem snapshots, the filesystem changeset can be used to present a series of image layers as if they were one cohesive filesystem.
+
+### Image JSON
+
+Each image has an associated JSON structure which describes some basic information about the image such as date created, author, and the ID of its parent image as well as execution/runtime configuration like its entrypoint, default arguments, CPU/memory shares, networking, and volumes.
+The JSON structure also references a cryptographic hash of each layer used by the image, and provides history information for those layers.
+This JSON is considered to be immutable, because changing it would change the computed ImageID.
```
Comment (wking) [inline config.md]: Can you add a new commit carrying #301, which changes `ImageID` → `[ImageID](#imageid)`?

File: config.md [hunk 19]
```diff
@@ -7,304 +7,170 @@ This specification outlines the JSON format describing images for use with a con
...
+`ChainID` is equal to the layer's `DiffID`.
+
+Otherwise the <code>ChainID</code> is given by the formula:
+<code>ChainID(layerN) = SHA256hex(ChainID(layerN-1) + " " + DiffID(layerN))</code>.
+
+### ImageID
+
+Each image's ID is given by the SHA256 hash of its configuration JSON.
+It is represented as a hexadecimal encoding of 256 bits, e.g., <code>sha256:a9561eb1b190625c9adb5a9513e72c4dedafc1cb2d4c5236c9a6957ec7dfd5a9</code>.
+Since the configuration JSON that gets hashed references hashes of each layer in the image, this formulation of the ImageID makes images content-addresable.
```
Comment (wking) [inline config.md]: Can you add a new commit carrying #301, which changes `configuration JSON` → [configuration JSON](#image-json), both on this line and the “Each image's ID…” line two above it?

**Comment by philips** [comment]
Addressed @coolljt0725 and @wking

File: config.md [hunk 20]
```diff
@@ -7,304 +7,171 @@ This specification outlines the JSON format describing images for use with a con
...
+    - **type** *string*, REQUIRED
+
+       MUST be set to `layers`.
+       Implementations MUST generate an error if they encounter a unknown value while verifying or unpacking an image.
+
+    - **diff_ids** *array*, REQUIRED
+
+       An array of layer content hashes (`DiffIDs`), in order from bottom-most to top-most.
+
+- **history** *array of object*, REQUIRED
```
Comment (wking) [inline config.md]: “array of object” → “array of objects”

File: config.md [hunk 21]
```diff
@@ -7,304 +7,171 @@ This specification outlines the JSON format describing images for use with a con
...
+   - **Cmd** *array of strings*, OPTIONAL
+
+     Default arguments to the entrypoint of the container.
+     These values act as defaults and are replaced with any specified when creating a container.
+     If an `Entrypoint` value is not specified, then the first entry of the `Cmd` array should be interpreted as the executable to run.
+
+   - **Volumes** *object*, OPTIONAL
+
+     A set of directories which should be created as data volumes in a container running this image. This field MAY be null.
+     If a file or folder exists within the image with the same path as a data volume, that file or folder is replaced with the data volume and is never merged. **NOTE:** This JSON structure value is unusual because it is a direct JSON serialization of the Go type `map[string]struct{}` and is represented in JSON as an object mapping its keys to an empty object.
```
Comment (wking) [inline config.md]: Split this into separate lines for each sentence.

File: config.md [hunk 22]
```diff
@@ -7,304 +7,171 @@ This specification outlines the JSON format describing images for use with a con
...
+`ChainID` is equal to the layer's `DiffID`.
+
+Otherwise the `ChainID` is given by the formula:
+`ChainID(layerN) = SHA256hex(ChainID(layerN-1) + " " + DiffID(layerN))`.
+
+### ImageID
+
+Each image's ID is given by the SHA256 hash of its configuration JSON.
+It is represented as a hexadecimal encoding of 256 bits, e.g., `sha256:a9561eb1b190625c9adb5a9513e72c4dedafc1cb2d4c5236c9a6957ec7dfd5a9`.
+Since the configuration JSON that gets hashed references hashes of each layer in the image, this formulation of the ImageID makes images content-addresable.
```
Comment (wking) [inline config.md]: This [still][1] needs “configuration JSON” links carried from #301.

[1]: https://github.com/opencontainers/image-spec/pull/333#discussion_r79770619

File: config.md [hunk 23]
```diff
@@ -7,304 +7,171 @@ This specification outlines the JSON format describing images for use with a con
...
+
+       The command which created the layer.
+
+    - **comment** *string*, OPTIONAL
+
+       A custom message set when creating the layer.
+
+    - **empty_layer** *boolean*, OPTIONAL
+
+       This field is used to mark if the history item created a filesystem diff, OPTIONAL
```
Comment (wking) [inline config.md]: “, OPTIONAL” → “.” (you say `empty_layer` is optional two lines above)

File: config.md [hunk 24]
```diff
@@ -7,304 +7,171 @@ This specification outlines the JSON format describing images for use with a con
...
+
+- **history** *array of object*, REQUIRED
+
+  Describes the history of each layer.
+  The array is ordered from bottom-most layer to top-most layer.
+  The object has the following fields:
+
+    - **created** *string*, OPTIONAL
+
+       Creation time, expressed as a ISO-8601 formatted combined date and time
```
Comment (wking) [inline config.md]: “a ISO” → “an ISO”.

File: config.md [hunk 25]
```diff
@@ -7,304 +7,171 @@ This specification outlines the JSON format describing images for use with a con
...
+
+   - **Env** *array of strings*, OPTIONAL
+
+     Entries are in the format of `VARNAME="var value"`.
+     These values act as defaults and are merged with any specified when creating a container.
+
+   - **Entrypoint** *array of strings*, OPTIONAL
+
+     A list of arguments to use as the command to execute when the container starts.
+     This value acts as a  default and is replaced by an entrypoint specified when creating a container. This field MAY be null.
```
Comment (wking) [inline config.md]: [Still][1] has a doubled space in “as a  default”.

[1]: https://github.com/opencontainers/image-spec/pull/333#discussion_r79762707

File: config.md (hunk 16 above)
Comment (coolljt0725) [inline config.md]: still lack of a `*` at the end of `object`. It looks like
`rootfs *object, REQUIRED`

**Comment by philips** [comment]
I think this is g2g @opencontainers/image-spec-maintainers

**Comment by vbatts** [comment]
Pretty huge overhaul.
/me :eyes:

**Comment by wking** [comment]
On Thu, Sep 22, 2016 at 12:01:30PM -0700, Brandon Philips wrote:
> I think this is g2g @opencontainers/image-spec-maintainers 

As of 85138c9 there are still a number of unaddressed comments on the
current diff:

* Carrying some links from #301 [1,2].
* Splitting to one sentence per line [3].
* Fixing a dangling ‘*object’ [4].
* Fixing a missing plural [5].
* Making ‘history’ OPTIONAL [6], since everything inside it is OPTIONAL.
* Fixing “a ISO” [7].

[1]: https://github.com/opencontainers/image-spec/pull/333/files#r79770351
[2]: https://github.com/opencontainers/image-spec/pull/333/files#r79976062
[3]: https://github.com/opencontainers/image-spec/pull/333/files#r79975924
[4]: https://github.com/opencontainers/image-spec/pull/333/files#r79766642
[5]: https://github.com/opencontainers/image-spec/pull/333/files#r79975858
[6]: https://github.com/opencontainers/image-spec/pull/333/files#r79981041
[7]: https://github.com/opencontainers/image-spec/pull/333/files#r79976240

**Comment by vbatts** [comment]
There are a couple of questions on the choices made for OPTIONAL/REQUIRED

**Comment by philips** [comment]
@vbatts what are the questions?

File: config.md [hunk 26]
```diff
@@ -7,304 +7,171 @@ This specification outlines the JSON format describing images for use with a con
...
-          </li>
-          <li>
-            <code>diff_ids</code> is an array of layer content hashes (<code>DiffIDs</code>), in order from bottom-most to top-most.
-          </li>
-        </ul>
-
-
-        Here is an example rootfs section:
-
-<pre>"rootfs": {
```
Comment (stevvooe) [inline config.md]: Are we dropping the examples?

Comment (wking) [inline config.md]: There's a big example at the end which gets verified vs. the JSON Schema.  I think that's sufficient.

**Comment by philips** [comment]
@wking I addressed all but the link from #301 in 974ac79. We can punt that discussion about adding readability links once this material change to the spec lands.

**Comment by jonboulle** [comment]
LGTM. Read over this and seems fine. It'll be much easier to land subsequent changes too.

[Approved with PullApprove](https://pullapprove.com/opencontainers/image-spec/pull-request/333/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=opencontainers/image-spec)

**Comment by vbatts** [comment]
Thanks for updating it.
LGTM

**Pull request merged**